### ISSUE TITLE
Returning correct scores for the entire sequence when sampling

### ISSUE THREAD

**Issue by jayelm** [OpenNMT/OpenNMT-py#1816]
My understanding is that when we call `DecodeStrategy.advance`, `log_probs` are normalized probabilities, i.e. they sum to 1 (since they have gone through a `generator` with a `LogSoftmax` op at the end):

https://github.com/OpenNMT/OpenNMT-py/blob/95aeefb4377617b94f94090d375eb8ae03dbd6c4/onmt/translate/greedy_search.py#L135-L151

But imagine if our sampling temperature is not 1 - when we call `self._pick` which calls `sample_with_temperature`, we modify the logits by that temperature when sampling:

https://github.com/OpenNMT/OpenNMT-py/blob/95aeefb4377617b94f94090d375eb8ae03dbd6c4/onmt/translate/greedy_search.py#L40-L56

When we do the `torch.div`, `logits` no longer sum to 1, so they are no longer valid log probabilities. Indeed, when we pass `logits` to `torch.distributions.Multinomial`, pytorch renormalizes the logits to do the sampling.

**However**, it seems like`topk_scores` is picked from the modified, unnormalized logits, which results in ill-defined scores?

Shouldn't the intended behavior be one of the following (though I'm not sure which):

1. Modify the logits for sampling, but return the original scores of the samples; so we should pick `topk_scores` from the original normalized `log_probs` that were passed to `sample_with_temperature`;
2. Return the temperature-modified scores, in which case we should renormalize the scores before we pick `topk_scores` (e.g. by using the normalized `dist.logits`).

**Comment by jayelm** [comment]
Ah, I realize it doesn't really make sense to renormalize if `keep_topk` is set, since if it were set to 1 (greedy decoding), we would just return log probs of 0 each time, but my question about the temperature stands.

**Comment by guillaumekln** [comment]
I guess you could renormalize but in general the returned scores are not expected to come from a normalized distribution. For example beam search returns the scores after penalties are applied (e.g. length penalty).

**Comment by jayelm** [comment]
Ok, thanks! One follow-up question: it seems like the scores returned from sampling are just scores for the *last* token, since `self.topk_scores` is overriden at each timestep:

https://github.com/OpenNMT/OpenNMT-py/blob/95aeefb4377617b94f94090d375eb8ae03dbd6c4/onmt/translate/greedy_search.py#L151

and these scores are read off only when `update_finished` is called:

https://github.com/OpenNMT/OpenNMT-py/blob/95aeefb4377617b94f94090d375eb8ae03dbd6c4/onmt/translate/greedy_search.py#L163-L169

Is this intended behavior? Is there a way I can get the score of an entire sampled sequence?

**Comment by guillaumekln** [comment]
> Is this intended behavior?

I don't think so. The scores of each step should probably be added. You could send a PR if you want to.

**Comment by vince62s** [comment]
fixed by #1958
